liblwgeom/g_serialized.c

```diff
@@ -284,8 +294,43 @@ int gserialized_cmp(const GSERIALIZED *g1, const GSERIALIZED *g2)
 {
 	int g1_is_empty, g2_is_empty, cmp;
 	GBOX box1, box2;
+	uint64_t hash1, hash2;
 	size_t sz1 = SIZE_GET(g1->size);
 	size_t sz2 = SIZE_GET(g2->size);
+	union floatuint x, y;
+
+	/*
+	* For two non-same planar points, we can skip a lot of machinery.
+	*/
+	if (
+		sz1 > 16 && // 16 is size of EMPTY, if it's larger - it has coordinates
+		sz2 > 16 &&
+		*(uint32_t*)(g1->data) == POINTTYPE &&
```

Comment:
The C Standard, 6.3.2.3, paragraph 7 [ISO/IEC 9899:2011], states
A pointer to an object or incomplete type may be converted to a pointer to a different object or incomplete type. If the resulting pointer is not correctly aligned for the referenced type, the behavior is undefined.

Reply:
The data buffer is constructed carefully so that aligned access is expected. Doubles exist on double alignment boundaries, and ints on their alignment boundary. https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/g_serialized.txt
Also, the storage type in pgSQL is defined as double aligned, so the segments we're getting back are on double-aligned boundares.
I'm not sure whether "it works" is sufficient for this objection.
One possible issue I guess is for geometries built on top of memory they allocated themselves, whether palloc() (when running within pgsql) or malloc() are returning double-aligned memory.

Reply:
There is no guarantee, than result of malloc have same alignment as double.

Reply:
And there is one more problem here, *(uint32_t*)(g1->data) and double *dptr = (double*)(g1->data); creates 2 aliases for same memory address with incompatible types.